Replace Jest with Vitest

jbouder opened the pull request:
This update is intended to provide the following:
- Replaces 8 Jest specific packages with 2 vitest related packages
- Same base unit test functionality
- Faster unit testing speeds
- Inline with existing Vite ecosystem

## Reference Issues or PRs

N/A

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [x] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

## Any other comments?
- Replaced Jest packages and configs with Vitest
- Updated all tests as needed

<!--
Please be aware that we are a loose team of volunteers, so patience is necessary;
assistance handling other issues is very welcome.
We value all user contributions. If we are slow to review, either the pull request needs some benchmarking, tinkering,
convincing, etc., or the reviewers are likely busy. In either case,
we ask for your understanding during the
review process.
Thanks for contributing to Nebari 🙏🏼!
-->

jbouder commented:
@aktech , can you check the vercel failure above? I thought I had access to Vercel, but can't seem to get to it.

aktech commented:
> I thought I had access to Vercel, but can't seem to get to it.

I think that's probably not possible in the free version IIRC.

Here are the logs:

<details>
<summary>Logs (click to expand)</summary>

```
[14:36:10.761] Running build in Washington, D.C., USA (East) – iad1
[14:36:10.996] Cloning github.com/nebari-dev/jhub-apps (Branch: vitest, Commit: 75d79a5)
[14:36:13.596] Cloning completed: 2.599s
[14:36:18.446] Restored build cache
[14:36:18.642] Running "vercel build"
[14:36:19.648] Vercel CLI 35.2.3
[14:36:22.154] Installing dependencies...
[14:36:26.023] 
[14:36:26.024] > nebari-hub@1.0.0 prepare
[14:36:26.025] > cd .. && husky
[14:36:26.025] 
[14:36:26.095] 
[14:36:26.095] added 13 packages, removed 12 packages, and changed 11 packages in 4s
[14:36:26.096] 
[14:36:26.096] 167 packages are looking for funding
[14:36:26.096]   run `npm fund` for details
[14:36:26.554] 
[14:36:26.554] > nebari-hub@1.0.0 build:dev
[14:36:26.554] > tsc && NODE_ENV=development vite build --mode development
[14:36:26.555] 
[14:36:39.322] [31mfailed to load config from /vercel/path0/ui/vite.config.ts[39m
[14:36:39.330] [31merror during build:
[14:36:39.331] Error: The package "@esbuild/linux-x64" could not be found, and is needed by esbuild.
[14:36:39.331] 
[14:36:39.331] If you are installing esbuild with npm, make sure that you don't specify the
[14:36:39.331] "--no-optional" or "--omit=optional" flags. The "optionalDependencies" feature
[14:36:39.332] of "package.json" is used by esbuild to install the correct binary executable
[14:36:39.332] for your current platform.
[14:36:39.332]     at generateBinPath (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1747:15)
[14:36:39.332]     at esbuildCommandAndArgs (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1817:33)
[14:36:39.332]     at ensureServiceIsRunning (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1974:25)
[14:36:39.332]     at build (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1873:26)
[14:36:39.332]     at bundleConfigFile (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:66435:24)
[14:36:39.337]     at loadConfigFromFile (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:66405:27)
[14:36:39.337]     at resolveConfig (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:66014:30)
[14:36:39.337]     at build (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:65111:24)
[14:36:39.337]     at CAC.<anonymous> (file:///vercel/path0/ui/node_modules/vite/dist/node/cli.js:828:11)[39m
[14:36:39.354] Error: Command "npm run build:dev" exited with 1
[14:36:39.999] 
```

</details>

jbouder commented:
> > I thought I had access to Vercel, but can't seem to get to it.
> 
> I think that's probably not possible in the free version IIRC.
> 
> Here are the logs:
> 
> Logs (click to expand)
> ```
> [14:36:10.761] Running build in Washington, D.C., USA (East) – iad1
> [14:36:10.996] Cloning github.com/nebari-dev/jhub-apps (Branch: vitest, Commit: 75d79a5)
> [14:36:13.596] Cloning completed: 2.599s
> [14:36:18.446] Restored build cache
> [14:36:18.642] Running "vercel build"
> [14:36:19.648] Vercel CLI 35.2.3
> [14:36:22.154] Installing dependencies...
> [14:36:26.023] 
> [14:36:26.024] > nebari-hub@1.0.0 prepare
> [14:36:26.025] > cd .. && husky
> [14:36:26.025] 
> [14:36:26.095] 
> [14:36:26.095] added 13 packages, removed 12 packages, and changed 11 packages in 4s
> [14:36:26.096] 
> [14:36:26.096] 167 packages are looking for funding
> [14:36:26.096]   run `npm fund` for details
> [14:36:26.554] 
> [14:36:26.554] > nebari-hub@1.0.0 build:dev
> [14:36:26.554] > tsc && NODE_ENV=development vite build --mode development
> [14:36:26.555] 
> [14:36:39.322] �[31mfailed to load config from /vercel/path0/ui/vite.config.ts�[39m
> [14:36:39.330] �[31merror during build:
> [14:36:39.331] Error: The package "@esbuild/linux-x64" could not be found, and is needed by esbuild.
> [14:36:39.331] 
> [14:36:39.331] If you are installing esbuild with npm, make sure that you don't specify the
> [14:36:39.331] "--no-optional" or "--omit=optional" flags. The "optionalDependencies" feature
> [14:36:39.332] of "package.json" is used by esbuild to install the correct binary executable
> [14:36:39.332] for your current platform.
> [14:36:39.332]     at generateBinPath (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1747:15)
> [14:36:39.332]     at esbuildCommandAndArgs (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1817:33)
> [14:36:39.332]     at ensureServiceIsRunning (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1974:25)
> [14:36:39.332]     at build (/vercel/path0/ui/node_modules/esbuild/lib/main.js:1873:26)
> [14:36:39.332]     at bundleConfigFile (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:66435:24)
> [14:36:39.337]     at loadConfigFromFile (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:66405:27)
> [14:36:39.337]     at resolveConfig (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:66014:30)
> [14:36:39.337]     at build (file:///vercel/path0/ui/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:65111:24)
> [14:36:39.337]     at CAC.<anonymous> (file:///vercel/path0/ui/node_modules/vite/dist/node/cli.js:828:11)�[39m
> [14:36:39.354] Error: Command "npm run build:dev" exited with 1
> [14:36:39.999] 
> ```

Added an optional dependency. Looks good now.

aktech commented on ui/src/utils/jupyterhub.test.ts:
```diff
@@ -22,10 +22,16 @@ import {
-  ...jest.requireActual('./jupyterhub'),
-  navigateToUrl: jest.fn(),
-}));
+vi.mock('./jupyterhub', async () => {
+  // Require the actual module to spread its properties
+  const actual = await vi.importActual('./jupyterhub');
+
+  return {
+    ...actual,
+    navigateToUrl: vi.fn(),
+  };
+});
+
```
Out of curiosity, do we have to make it async to make the test work? I thought Vitest is a drop-in replacement for Jest

jbouder replied:
Looks like `vi.importActual` returns a promise, so it would have to either be async, or would need to chain with `then` to handle what happens when the promise is returned.

That said, in general, vite and jest provide a nearly identical implementation, but not 100%.

aktech replied:
I see, thanks for the explanation.
